CORDA-4105 Add public API to allow custom serialization schemes

williamvigorr3 opened the pull request:
Added public interface CustomSerializationScheme so custom serialization schemes can be implemented. Added adapter code which takes an implementation of this interface and creates an implementation of SerializationScheme (this is the internal equivalent which existed before this PR). Currently this adapter is only used by tests, the idea is to eventually use this in the node after discovering implementations of CustomSerializationScheme.

williamvigorr3 committed:
CORDA-4105 Add public API to allow custom serialization schemes

williamvigorr3 committed:
Fix Detekt

williamvigorr3 committed:
Suppress warning

williamvigorr3 committed:
Fix usused import

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,39 @@
+    fun getSerializationMagic(): CustomSerializationMagic
+
+    /**
+     * This method must deserialize (any) object from SerializedBytes.
```
Worth saying what the input parameters are and what the output is (for this and all other public methods).

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,39 @@
+    fun serialize(obj: Any, context: CustomSerializationContext): SerializedBytes<*>
+}
+
+class CustomSerializationMagic(val magicNumber: Int)
```
Needs a kdoc to describe what this is.

mnesbit replied:
Probably also have to be clear on threading requirement

williamvigorr3 replied:
I have added detailed Kdoc including threading requirements (which is something that would be very easy to get wrong).

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,39 @@
+
+class CustomSerializationMagic(val magicNumber: Int)
+
+interface CustomSerializationContext {
```
This too.

JamesHR3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/CustomSerializationSchemeAdapter.kt:
```diff
@@ -0,0 +1,49 @@
+
+    companion object {
+        //If only this was C
+        const val SIZE_OF_INT = 4
```
I think we should call this something like `SERIALIZATION_MAGIC_BUFFER_SIZE` - something that describes what it is used for.

williamvigorr3 replied:
Changed it to `SERIALIZATION_SCHEME_ID_SIZE`

JamesHR3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/CustomSerializationSchemeAdapter.kt:
```diff
@@ -0,0 +1,49 @@
+            throw NotSerializableException("Scheme ${customScheme::class.java} is incompatible with blob." +
+                    " Magic from blob = $readMagic (Expected = $serializationSchemeMagic)")
+        val withOutMagic = byteSequence.bytes.slice(serializationSchemeMagic.size .. byteSequence.size - 1).toByteArray()
+      @Suppress("UNCHECKED_CAST")
```
This cast is technically unsafe right? Should we verify that we get a `clazz` instance back from the custom serializer and throw if we don't?

williamvigorr3 replied:
This is no longer needed. After properly templating the API.

rick-r3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * This must return a magic number used to uniquely (within a network) identify the Scheme.
+     */
+    fun getSerializationMagic(): CustomSerializationMagic
```
Shall we perhaps just make this `SchemeId` and make it an `Int`, and avoid use of "magic" in the customer facing API ?

i.e. `fun getSchemeId(): Int`

mnesbit replied:
I presume whatever they pick won't clash with any existing magic bytes?

williamvigorr3 replied:
I have changed it to an int. It can't clash as the adapter class encodes "CUS" and then the `SchemeId` as a 4 byte int. For AMQP we encode  "CORDA" and then 0 (as a byte) and then 1 (as a byte).

rick-r3 reviewed (commented):
Simplification and naming suggestion.

williamvigorr3 committed:
Improve API to use generics

This does not break Java support (only Intelij gets confused).

williamvigorr3 committed:
Add more detailed documentation to public interfaces

williamvigorr3 committed:
Change internal variable name after rename public API

rick-r3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,51 @@
+     * @param context used to pass information about how the object should be deserialized.
+     */
+    @Throws(NotSerializableException::class)
+    fun <T : Any> deserialize(bytes: SerializedBytes<T>, clazz: Class<T>, context: CustomSerializationContext): T
```
I'm having second thoughts about these signatures.  I think we shouldn't use `SerializedBytes` here or in `serialize` because by definition these are missing the magic bytes header that all `SerializedBytes` have.  Can we switch these to `ByteSequence` ?  I know that wipes off T, but I think it is worth it.

Otherwise things might well get confusing.

williamvigorr3 replied:
I managed to do this and no casts were needed in the end.

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/CustomSerializationSchemeAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+        if (readMagic != serializationSchemeMagic)
+            throw NotSerializableException("Scheme ${customScheme::class.java} is incompatible with blob." +
+                    " Magic from blob = $readMagic (Expected = $serializationSchemeMagic)")
+        val withOutMagic = byteSequence.bytes.slice(serializationSchemeMagic.size until byteSequence.size).toByteArray()
```
See my other comment about not using `SerializedBytes` on the new API.  Note: `ByteSequence.subSequence()` will help here once not using `SerializedBytes`.

williamvigorr3 committed:
Update Public API to use ByteSquence instead of SerializedBytes

williamvigorr3 committed:
Remove unused import

williamvigorr3 committed:
Fix whitespace.

JamesHR3 commented on core/src/main/kotlin/net/corda/core/serialization/CustomSerializationScheme.kt:
```diff
@@ -0,0 +1,52 @@
+ * [CustomSerializationScheme.deserialize] or [CustomSerializationScheme.serialize] are called internally.
+ */
+@DoNotImplement
+interface CustomSerializationContext {
```
Are the values provided here options for (de)serialization? I wonder if it's worth at least pointing out a sensible default if the user has no specific requirements for these things. (I'm sure there's an "allow everything" under `ClassWhitelist` for example.) Incidentally, is `ClassWhitelist` public?

I notice `@DoNotImplement` here too - if this shouldn't be implemented, might be worth putting it in a separate file so it's less confusing what's actually part of the API and what isn't? (Might be worth doing this anyway.) If it should be implemented, how is the context and the scheme linked internally?

williamvigorr3 replied:
So the idea here is we implement this interface inside Corda and a user should only use it inside there implementation of `CustomSerializationScheme.deserialize` and `CustomSerializationScheme.serialize`. I would still say it is still a part of the public API, just that a user should use it not implement it directly (like FlowSession for example).

The classloader might be required by the serialization library. If it calls class.forname after adding the name of the class to serialized blob.

I think the class whitelist is built from things annotated with `@CordaSerializable` (I will check that this is actually correct). These things can be both internal in Corda or user annotated in CorDapp itself. So a secure implementation will check that a class is in the whitelist before constructing it. Maybe this check is not necessary all the time it depends on the when deserialize is called.

williamvigorr3 replied:
`ClassWhitelist` is public already.

JamesHR3 replied:
Right, got it - we'll give you a `CustomSerializationContext`, which is a `SerializationContext` but with just those fields we want you to inspect. We're guaranteed to give you the "right" classloader/whitelist, you just get the opportunity to look at those if you want to.

I think the following things might help a bit:
- Rename `CustomSerializationContext` (I'm not sure what to). I think the `Custom` out the front implies this is something you might want to implement (customize) to change the (de)serialization behaviour, which is not the case.
- Move it to its own file - that keeps the things you want to implement separate from the things you don't.
- Add to the kdocs some information about why you might want to inspect these properties on the context (security when constructing types on the whitelist, to ensure you can load the right type with `Class.forName` on the classloader)

williamvigorr3 replied:
Okay I renamed this to `SerializationSchemeContext`. If moved it to it's own file and added more KDOC explaining how it is supposed to be used.

williamvigorr3 committed:
Add added public API to .ci/api-current.txt

mnesbit commented on .ci/api-current.txt:
```diff
@@ -5935,6 +5935,20 @@ public @interface net.corda.core.serialization.CordaSerializationTransformRename
+  public abstract int getSchemeId()
+  @NotNull
+  public abstract net.corda.core.utilities.ByteSequence serialize(T, net.corda.core.serialization.CustomSerializationContext)
+##
```
This looks fine and the usual false flag that we get when abstract methods are added to corda

williamvigorr3 committed:
Improve public interfaces

Rename CustomSchemeContext to SerializationSchemeContext to improve
clarity and move to it's own file. Improve kdoc to make things less
confusing.

JamesHR3 reviewed (dismissed):
LGTM! Think the kdocs look a lot better now.

williamvigorr3 committed:
Update API current with changed API
